src/mlpack/core/optimizers/pso/pso_impl.hpp

```diff
@@ -0,0 +1,137 @@
+/**
+ * @file pso_impl.hpp
+ * @author Adeel Ahmad
+ *
+ * Implementation of the Particle Swarm Optimizer as proposed
+ * by J. Kennedy et al. in "Particle swarm optimization".
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PSO_PSO_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_PSO_PSO_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "pso.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename VelocityVectorType>
+PSOType<VelocityVectorType>::PSOType(const size_t lambda,
+                                  const size_t dimension,
+                                  const double interiaWeight,
+                                  const double cognitiveAcceleration,
+                                  const double socialAcceleration,
+                                  const double velocityLowerBound,
+                                  const double velocityUpperBound,
+                                  const double postionLowerBound,
+                                  const double postionUpperBound,
+                                  const size_t maxIterations,
+                                  const double tolerance,
+                                  const VelocityVectorType& velocityType) :
+    lambda(lambda),
+    dimension(dimension),
+    interiaWeight(interiaWeight),
+    cognitiveAcceleration(cognitiveAcceleration),
+    socialAcceleration(socialAcceleration),
+    velocityLowerBound(velocityLowerBound),
+    velocityUpperBound(velocityUpperBound),
+    postionLowerBound(postionLowerBound),
+    postionUpperBound(postionUpperBound),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    velocityType(velocityType)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename VelocityVectorType>
+template<typename DecomposableFunctionType>
+double PSOType<VelocityVectorType>::Optimize(
+    DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Randomly initialize the particle position and velocity.
+  // Following a heuristic, the swarm size is set to 2 * dimension.
+  lambda = 2 * dimension;
+  particlePosition.randu(dimension);
```

Comment:
We should use the iterate dimension here, which should contain the initial starting point. I think the idea here is that ``dimension`` defines the population size, correct me if I'm wrong. So I think it should look something like:

```c++
particlePosition.randu(iterate.n_rows, iterate.n_cols);
```

Which defines a single particle, I guess an easy option would be to use ``arma::cube``  to store populations. But depending on the population and function size this could be expensive. So another option would be to use ``arma::cube`` that stores only the current best solution and the solution we are testing:

```c++
arma::cube particlePosition = randu<arma::cube>(iterate.n_rows, iterate.n_cols, 2);
```

and in the evaluation step, which currently only uses a single particle we do something like:

```c++
for (size_t p = 0; p < populationSize; ++p)
{
  currentObjective = function.Evaluate(particlePosition.slice(0));
  if (currentObjective < lastObjectiveIndividual)
  {
   // slice(0) contains the current working particle and slice(1) the current best particle.
    particlePosition.slice(1) = particlePosition.slice(0);    
  }
  ... 
}
```

We should be able to do something clever to avoid the extra copy, but we can do that later. I hope something I said is helpful, let me know if I should clarify anything.